module/zfs/arc.c

```diff
@@ -1702,6 +1750,156 @@ arc_buf_try_copy_decompressed_data(arc_buf_t *buf)
 }
 
 /*
+ * Return the size of the block, b_pabd, that is stored in the arc_buf_hdr_t.
+ */
+static uint64_t
+arc_hdr_size(arc_buf_hdr_t *hdr)
+{
+	uint64_t size;
+
+	if (arc_hdr_get_compress(hdr) != ZIO_COMPRESS_OFF &&
+	    HDR_GET_PSIZE(hdr) > 0) {
+		size = HDR_GET_PSIZE(hdr);
+	} else {
+		ASSERT3U(HDR_GET_LSIZE(hdr), !=, 0);
+		size = HDR_GET_LSIZE(hdr);
+	}
+	return (size);
+}
+
+/*
+ * This function will take a header that only has raw encrypted data in
+ * b_crypt_hdr.b_rabd and decrypts it into a new buffer which is stored in
+ * b_l1hdr.b_pabd. If designated on the header, this function will also
+ * decompress the data as well.
+ */
+static int
+arc_hdr_decrypt(arc_buf_hdr_t *hdr, kmutex_t *hash_lock, spa_t *spa,
+    uint64_t dsobj)
+{
+	int ret;
+	dsl_crypto_key_t *dck = NULL;
+	abd_t *cabd = NULL;
+	void *tmp = NULL;
+
+	if (hash_lock != NULL)
+		mutex_enter(hash_lock);
+
+	/*
+	 * Check that we only have an encrypted copy of the data. This was
+	 * already checked in arc_buf_fill() as a quick check, but we do it
+	 * again now under the hash_lock to make sure nothing has changed.
+	 * If this isn't true there is no work to do so we can simply return.
+	 */
+	if (!HDR_HAS_RABD(hdr) || hdr->b_l1hdr.b_pabd != NULL)
+		goto out_unlock;
+
+	arc_hdr_alloc_abd(hdr, B_FALSE);
+
+	/*
+	 * We must be careful to use the passed-in dsobj value here and
+	 * not the value in b_dsobj. b_dsobj is meant to be a best guess for
+	 * the L2ARC, which has the luxury of being able to fail without real
+	 * consequences (the data simply won't make it to the L2ARC). In
+	 * reality, the dsobj stored in the header may belong to a dataset
+	 * that has been unmounted or otherwise disowned, meaning the key
+	 * won't be accessible via that dsobj anymore.
+	 */
+	ret = spa_keystore_lookup_key(spa, dsobj, FTAG, &dck);
+	if (ret != 0)
+		goto error;
+
+	ret = zio_do_crypt_abd(B_FALSE, &dck->dck_key,
+	    hdr->b_crypt_hdr.b_salt, hdr->b_crypt_hdr.b_ot,
+	    hdr->b_crypt_hdr.b_iv, hdr->b_crypt_hdr.b_mac,
+	    HDR_GET_PSIZE(hdr), hdr->b_l1hdr.b_pabd,
+	    hdr->b_crypt_hdr.b_rabd);
+	if (ret == ZIO_NO_ENCRYPTION_NEEDED) {
+		abd_copy(hdr->b_l1hdr.b_pabd, hdr->b_crypt_hdr.b_rabd,
+		    HDR_GET_PSIZE(hdr));
+	} else if (ret != 0) {
+		goto error;
+	}
+
+	if (HDR_GET_COMPRESS(hdr) != ZIO_COMPRESS_OFF &&
+	    !HDR_COMPRESSION_ENABLED(hdr)) {
+		/*
+		 * We want to make sure that we are correctly honoring the
+		 * zfs_abd_scatter_enabled setting, so we allocate an abd here
+		 * and then loan a buffer from it, rather than allocating a
+		 * linear buffer and wrapping it in an abd later.
+		 */
+		cabd = arc_get_data_abd(hdr, arc_hdr_size(hdr), hdr);
+		tmp = abd_borrow_buf(cabd, arc_hdr_size(hdr));
+
+		ret = zio_decompress_data(HDR_GET_COMPRESS(hdr),
+		    hdr->b_l1hdr.b_pabd, tmp, HDR_GET_PSIZE(hdr),
```

Comment:
My only issue with that is that then we will not be honoring `zfs_abd_scatter_enabled`. I do not think this is an incredibly performance-sensitive path either, since this will realistically only be run in 2 circumstances: decrypting dnode blocks and reading data that is already in the ARC due to a scrub.

Let me know what you think and I can change it if you still feel it is appropriate.